src/kvstore/raftex/RaftPart.cpp

```diff
@@ -1345,29 +1355,34 @@ void RaftPart::processAskForVoteRequest(const cpp2::AskForVoteRequest& req,
     resp.set_error_code(cpp2::ErrorCode::E_TERM_OUT_OF_DATE);
     return;
   }
+
+  // Ok, no reason to refuse, we will vote for the candidate
+  LOG(INFO) << idStr_ << "The partition will vote for the candidate " << candidate
+            << ", isPreVote = " << req.get_is_pre_vote();
+
   if (req.get_is_pre_vote()) {
     // return succeed if it is prevote, do not change any state
     resp.set_error_code(cpp2::ErrorCode::SUCCEEDED);
     return;
   }
 
-  // Ok, no reason to refuse, we will vote for the candidate
-  LOG(INFO) << idStr_ << "The partition will vote for the candidate " << candidate
-            << ", isPreVote = " << req.get_is_pre_vote();
-
-  // Before change role from leader to follower, check the logs locally.
-  if (role_ == Role::LEADER && wal_->lastLogId() > lastLogId_) {
+  // not a pre-vote, need to rollback wal if necessary
+  // role_ and term_ has been set above
+  if (oldRole == Role::LEADER && wal_->lastLogId() > lastLogId_) {
```

Comment:
Still not very clear about this part, let me try to clarify:
1. is `lastLogId_` == commitIndex?
2. wal->lastLogId() returns entries appended but might not have been commited?

if that's the case, then two questions:

1. `wal_->rollbackToLog(lastLogId_)` means to rolling back entries that haven't been commited yet?
2. is it possible that the entries not commited have been replicated among the majority of the raft peers?

Reply:
It is hard to clarify, in short version:
    * As for leader, `lastLogId_` is `commitIndex` in paper
    * As for follower, it is not

So rollback means to rolling back entries that haven't been committed yet. And yes, the logs which may have been replicated to majority. What will happen is that, no matter we rollback or not, we will report error to client. So whether the log will be applied, depends on new leader.

Reply:
I know why we should roll back here but still kind of warry about the potential impact of this part though I cann't tell what the problem is. I'll do more test.